arelle/plugin/validate/ROS/rules/ros.py

```diff
@@ -337,27 +338,48 @@ def rule_ros20(
     used for the majority of monetary facts.
     """
     principal_currency_facts = val.modelXbrl.factsByLocalName.get(PRINCIPAL_CURRENCY, set())
-    principal_currency_values = set(fact.text for fact in principal_currency_facts)
+    principal_currency_values = {
+        currencyDimensionCode
+        for pc_fact in principal_currency_facts
+        if (currencyDimensionCode := _getCurrencyDimensionCode(val.modelXbrl, pc_fact))
+    }
     if len(principal_currency_values) != 1:
         yield Validation.error(
             "ROS.20",
             _("'PrincipalCurrencyUsedInBusinessReport' must exist and have a single value.  Values found: %(principal_currency_values)s."),
             modelObject=principal_currency_facts,
-            principal_currency_values=principal_currency_values,
+            principal_currency_values=sorted(principal_currency_values),
         )
     else:
         principal_currency_value = principal_currency_values.pop()
         monetary_facts = list(val.modelXbrl.factsByDatatype(False, qnXbrliMonetaryItemType))
-        monetary_units = [list(fact.utrEntries)[0].unitName for fact in monetary_facts if fact.unit is not None]
+        monetary_units = [list(fact.utrEntries)[0].unitId for fact in monetary_facts if fact.unit is not None]
         unit_counts = Counter(monetary_units)
         principal_currency_value_count = unit_counts[principal_currency_value]
-        for unit, count in unit_counts.items():
+        for count in unit_counts.values():
             if count > principal_currency_value_count:
                 yield Validation.warning(
                     "ROS.20",
-                    _("'PrincipalCurrencyUsedInBusinessReport' has a value of %(principal_currency_value)s, "
+                    _("'PrincipalCurrencyUsedInBusinessReport' has a %(currencies_dimension)s value of %(principal_currency_value)s, "
                       "which must match the functional(majority) unit of the financial statement."),
                     modelObject=principal_currency_facts,
+                    CurrenciesDimension=CURRENCIES_DIMENSION,
                     principal_currency_value=principal_currency_value,
                 )
                 break
+
+def _getCurrencyDimensionCode(modelXbrl: ModelXbrl, fact: ModelInlineFact) -> str | None:
+    if fact.context is None:
+        return None
+    for dim, mem in fact.context.qnameDims.items():
+        if dim.localName != CURRENCIES_DIMENSION:
+            continue
+        if mem.xValid < VALID:
+            return None
+        mem_concept = modelXbrl.qnameConcepts.get(mem.xValue)
+        if mem_concept is None:
+            return None
+        for ref_rel in modelXbrl.relationshipSet(XbrlConst.conceptReference).fromModelObject(mem_concept):
+            if code := ref_rel.toModelObject.findtext("{*}Code"):
```

Comment:
It's somewhat loose to match without a namespace.  Perhaps a small comment saying that this is intended to match `uk-ref:Code`?